content/blog/2023/2024-03-15-getting-started-with-spring-security-and-jwt.md

```diff
@@ -0,0 +1,965 @@
+ | **Data**                       | Basic Authentication is meant to handle only credentials (typically username-password). | JWT can include additional information such as id, roles, etc. Once the signature is validated, the server can trust the data sent by the client thus avoiding any additional lookups that maybe needed otherwise.|  
+
+
+## Implementing JWT in a Spring Boot Application
+Now that we understand JWT better, let's implement it in a simple Spring Boot application.
+In our *pom.xml*, let's add the below dependencies:
+
+````xml
+<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-web</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-security</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>io.jsonwebtoken</groupId>
+			<artifactId>jjwt-api</artifactId>
+			<version>0.11.1</version>
+		</dependency>
+		<dependency>
+			<groupId>io.jsonwebtoken</groupId>
+			<artifactId>jjwt-impl</artifactId>
+			<version>0.11.1</version>
+			<scope>runtime</scope>
+		</dependency>
+		<dependency>
+			<groupId>io.jsonwebtoken</groupId>
+			<artifactId>jjwt-jackson</artifactId>
+			<version>0.11.1</version>
+			<scope>runtime</scope>
+		</dependency>
+````
+We have created a simple spring boot Library application that uses an in-memory H2 database to store data.
+The application is configured to run on port 8083. To run the application:
+
+````text
+mvnw clean verify spring-boot:run (for Windows)
+./mvnw clean verify spring-boot:run (for Linux)
+````
+### Intercepting the Spring Security Filter Chain for JWT
+
+The application has a REST endpoint `/library/books/all` to get all the books stored in the DB. IF we make this GET call via Postman, we get a `401 UnAuthorized` error:
+{{% image alt="settings" src="images/posts/spring-security-and-jwt/Postman401.png" %}}
+This is because, **the `spring-boot-starter-security` dependency added in our *pom.xml*, automatically brings in Basic authentication to all the endpoints created.**
+Since we haven't specified any credentials in Postman we get the `UnAuthorized` error.
+For the purpose of this article, we need to replace Basic Authentication with JWT-based authentication.
+We know that Spring provides security to our endpoints, by triggering a chain of filters that handle authentication and authorization for every
+request. The [`UsernamePasswordAuthenticationFilter`](https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/web/authentication/UsernamePasswordAuthenticationFilter.html) is responsible for validating the credentials for every request. 
+
+Let's create a new `Filter` called `JwtFilter` and add it before `UsernamePasswordAuthenticationFilter` in the filter chain. This will make sure that `JwtFilter` will be used to validate the token sent in the `Authorization` header along with the request.
```

Comment:
I think we are missing some code snippets. We talk about the `JwtFilter` class that we want to create and insert into the filter chain. However, there's no code for these two things. While reading, this creates quite a jump. We talk about the `JwtFilter` class that we want to create and then show the `createToken` method.

Reply:
I have remove the "and add it before the filter chain" part as I have already covered that in the further section.
Also added the JwtFilter code snippet for the user to corelate